src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -0,0 +1,100 @@
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.HttpOverStreams.HttpContent;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.HttpOverStreams
+{
+    internal class DatadogHttpClient
+    {
+        private const int BufferSize = 10240;
+        private static readonly Vendors.Serilog.ILogger Logger = DatadogLogging.For<DatadogHttpClient>();
+
+        public Task<HttpResponse> SendAsync(HttpRequest request, Stream requestStream, Stream responseStream, CancellationToken cancellationToken)
+        {
+            Task.Run(() => SendRequest(request, requestStream, cancellationToken), cancellationToken);
+            return ReadResponse(responseStream, cancellationToken);
+        }
+
+        private static async Task SendRequest(HttpRequest request, Stream requestStream, CancellationToken cancellationToken)
+        {
+            using (var writer = new StreamWriter(requestStream, Encoding.ASCII, BufferSize, leaveOpen: true))
+            {
+                await DatadogHttpHeaderHelper.WriteLeadingHeaders(request, writer).ConfigureAwait(false);
+
+                foreach (var header in request.Headers)
+                {
+                    await DatadogHttpHeaderHelper.WriteHeader(writer, header).ConfigureAwait(false);
+                }
+
+                await DatadogHttpHeaderHelper.WriteEndOfHeaders(writer).ConfigureAwait(false);
+            }
+
+            await request.Content.CopyToAsync(requestStream).ConfigureAwait(false);
+            Logger.Debug("Datadog HTTP: Flushing stream.");
+            await requestStream.FlushAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<HttpResponse> ReadResponse(Stream responseStream, CancellationToken cancellationToken)
+        {
+            var headers = new HttpHeaders();
+            int statusCode = 0;
+            string responseMessage = null;
+
+            // hack: buffer the entire response so we can seek
+            var memoryStream = new MemoryStream();
+            responseStream.CopyTo(memoryStream);
+            memoryStream.Position = 0;
+            int streamPosition = 0;
+
+            using (var reader = new StreamReader(memoryStream, Encoding.ASCII, detectEncodingFromByteOrderMarks: false, BufferSize, leaveOpen: true))
```

Comment:
Have you checked the agent is indeed using ASCII encoding? Maybe we should test what happens when we put accents in service names

Reply:
This is the encoding for the http headers, which is always(?) ASCII. Service names are in the contents which has its own encoding. In our case the content is always JSON (which is requires by spec to be UTF-8), but the serializer handles that, not us.